src/mod/endpoints/mod_pulseaudio/pablio.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * $Id: pablio.c 1151 2006-11-29 02:11:16Z leland_lucius $
+ * pablio.c
+ * PulseAudio abstraction
+ *
+ * Contributor(s):
+ *
+ * Phil Burk, http://www.softsynth.com
+ * Jérôme Poulin <jeromepoulin@gmail.com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files
+ * (the "Software"), to deal in the Software without restriction,
+ * including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
+ * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <switch.h>
+#include <stdlib.h>
+#include <string.h>
+#include <pulse/timeval.h>
+#include <pulse/simple.h>
+#include "pablio.h"
+
+long WriteAudioStream(PABLIO_Stream * aStream, void *data, size_t datalen, switch_timer_t *timer)
+{
+	pa_simple_write(aStream->ostream, data, datalen, NULL);
+	switch_core_timer_next(timer);
```

Comment:
```suggestion
```
Why use a timer? This function gets called inside the `channel_write_frame` callback, that is supposed synchronized.

Reply:
Anything in pablio.h/c is code that was imported from PortAudio, so it might just be remains from the ring buffer that was there.

I didn't take time to understand this part, I'll give it a try without the timer, thanks for the suggestion.

I'm also thinking about removing or restructuring/renaming pablio.h/c to more accurately represent PulseAudio's API.

Reply:
Thank you for replying.

I tried a call without the timers and everything seems to work. 

Another issue I'm facing regards latency. I'm using this mod on an embedded device with low computational resources. When the `module-echo-cancel` is enabled the latency grows heavily (several seconds).
Using the `module-loopback` together with `module-echo-cancel` does not present this issue: latency still increases but only of several tenths of a second. So I guess the Simple API is not suited for low-latency applications.
Would you have any suggestion about this? 

Thank you

Reply:
When I decided about using libpulse-simple library instead of the fully fledged libpulse here is what I took into consideration.

I took mod_portaudio as a base for building upon since it was more recent than mod_alsa, in mod_portaudio, pablio.c's main job was to build a ring buffer inside freeswitch and to make calls to ReadAudioStream/WriteAudioStream synchronous. This is also why the pablio files are so lightweight now.

The main goal of using libpulse is to be able to attach pulsaudio's library to an existing main loop and make those calls asynchronous.  The other advantage of using libpulse is to have control over audio volumes and other internals to better handle pulseaudio specific events.

The pulse-simple library is a wrapper around libpulse that automatically creates a thread and a mainloop for pulseaudio and handle all the locking and callbacks to make sure that read/write calls are synchronous.

Now, the part I'm not sure is whatever Freeswitch can work with asynchronous read/write for audio channels. From a design perspective it'd make sense to remove the pulseaudio mainloop thread as this would be less overhead, and it seems like quite a big rewrite.

One other detail, I just noticed latency_msec is hard-coded to 40, I was sure it was configurable, you could try making this a configuration and tweaking it for your application, it might help.

Reply:
I totally agree with you, it'd make sense to remove PulseAudio's main loop, but as you said it seems quite complicated. 

What do you mean by asynchronous read/write with Freeswitch? As far as I know, audio is read/written with the IO routines `channel_read_frame` and `channel_write_frame`. Is this already an asynchronous behavior, isn't it? 

Anyway, after several tests, the threaded main loop of PulseAudio does not seem too heavy by itself, but when the echo cancellation is enabled server-side the overhead becomes too much for low computational resources, as my desktop can make calls without delay while the low-powered device cannot.

I wonder if the overall logic could be lighten in some way.